File: frontend/lib/data/models/goal.dart
import 'dashboard_summary.dart';
import 'expense_category.dart';

/// Mirrors the backend's `GoalType` enum values byte-for-byte.
enum GoalType {
  savings,
  spendingCap;

  String get wireValue => this == GoalType.savings ? 'savings' : 'spending_cap';

  static GoalType fromJson(String value) => value == 'spending_cap' ? GoalType.spendingCap : GoalType.savings;
}

/// Computed server-side on every read (backend app/services/goal_forecast.py)
/// — the client never derives a verdict itself.
enum GoalVerdict {
  onTrack('on_track', 'On track'),
  atRisk('at_risk', 'At risk'),
  offTrack('off_track', 'Off track'),
  achieved('achieved', 'Achieved'),
  missed('missed', 'Missed'),
  insufficientData('insufficient_data', 'Too early to tell');

  final String wireValue;
  final String label;

  const GoalVerdict(this.wireValue, this.label);

  static GoalVerdict fromJson(String value) {
    return GoalVerdict.values.firstWhere((v) => v.wireValue == value, orElse: () => GoalVerdict.insufficientData);
  }
}

double? _nullableAmount(dynamic value) => value == null ? null : parseAmount(value);

class GoalForecast {
  final GoalVerdict verdict;
  final double progressKrw;
  final double progressPercent;
  final int daysElapsed;
  final int daysTotal;
  final int daysRemaining;
  final double paceKrwPerDay;
  final double? projectedFinalKrw;
  final double? requiredKrwPerDay;
  final double? requiredChangeKrwPerDay;

  /// Same shape as the Clarity Score's factors — the XAI reasoning behind
  /// [verdict]. Never render the verdict without these.
  final List<ClarityFactor> factors;

  const GoalForecast({
    required this.verdict,
    required this.progressKrw,
    required this.progressPercent,
    required this.daysElapsed,
    required this.daysTotal,
    required this.daysRemaining,
    required this.paceKrwPerDay,
    required this.projectedFinalKrw,
    required this.requiredKrwPerDay,
    required this.requiredChangeKrwPerDay,
    required this.factors,
  });

  factory GoalForecast.fromJson(Map<String, dynamic> json) {
    return GoalForecast(
      verdict: GoalVerdict.fromJson(json['verdict'] as String),
      progressKrw: parseAmount(json['progress_krw']),
      progressPercent: parseAmount(json['progress_percent']),
      daysElapsed: json['days_elapsed'] as int,
      daysTotal: json['days_total'] as int,
      daysRemaining: json['days_remaining'] as int,
      paceKrwPerDay: parseAmount(json['pace_krw_per_day']),
      projectedFinalKrw: _nullableAmount(json['projected_final_krw']),
      requiredKrwPerDay: _nullableAmount(json['required_krw_per_day']),
      requiredChangeKrwPerDay: _nullableAmount(json['required_change_krw_per_day']),
      factors: (json['factors'] as List).map((f) => ClarityFactor.fromJson(f as Map<String, dynamic>)).toList(),
    );
  }
}

class Goal {
  final String id;
  final String name;
  final GoalType goalType;
  final double targetAmountKrw;
  final double startingAmountKrw;
  final ExpenseCategory? category;
  final DateTime startDate;
  final DateTime targetDate;
  final GoalForecast forecast;

  const Goal({
    required this.id,
    required this.name,
    required this.goalType,
    required this.targetAmountKrw,
    required this.startingAmountKrw,
    required this.category,
    required this.startDate,
    required this.targetDate,
    required this.forecast,
  });

  factory Goal.fromJson(Map<String, dynamic> json) {
    return Goal(
      id: json['id'] as String,
      name: json['name'] as String,
      goalType: GoalType.fromJson(json['goal_type'] as String),
      targetAmountKrw: parseAmount(json['target_amount_krw']),
      startingAmountKrw: parseAmount(json['starting_amount_krw']),
      category: json['category'] == null ? null : ExpenseCategory.fromJson(json['category'] as String),
      startDate: DateTime.parse(json['start_date'] as String),
      targetDate: DateTime.parse(json['target_date'] as String),
      forecast: GoalForecast.fromJson(json['forecast'] as Map<String, dynamic>),
    );
  }
}

File: backend/app/schemas/goal.py
import enum
import uuid
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.models.enums import ExpenseCategory, GoalType
from app.schemas.common import XAIFactor


class GoalVerdict(str, enum.Enum):
    """Computed on every read from real contributions / expenses, never stored."""

    ON_TRACK = "on_track"
    AT_RISK = "at_risk"
    OFF_TRACK = "off_track"
    ACHIEVED = "achieved"
    MISSED = "missed"
    INSUFFICIENT_DATA = "insufficient_data"


class GoalCreate(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    goal_type: GoalType
    target_amount_krw: Decimal = Field(gt=0, max_digits=12, decimal_places=2)
    starting_amount_krw: Decimal = Field(default=Decimal("0"), ge=0, max_digits=12, decimal_places=2)
    category: ExpenseCategory | None = None
    # Omitted = today (resolved in the service layer).
    start_date: date | None = None
    target_date: date

    @model_validator(mode="after")
    def _consistent_with_goal_type(self):
        start = self.start_date or date.today()
        if self.target_date < start:
            raise ValueError("target_date must be on or after start_date")
        if self.goal_type == GoalType.SAVINGS and self.category is not None:
            raise ValueError("category only applies to spending_cap goals")
        if self.goal_type == GoalType.SPENDING_CAP and self.starting_amount_krw != 0:
            raise ValueError("starting_amount_krw only applies to savings goals")
        return self


class GoalContributionCreate(BaseModel):
    # Negative = a withdrawal from the savings pot.
    amount_krw: Decimal = Field(max_digits=12, decimal_places=2)
    contributed_on: date | None = None
    note: str | None = Field(default=None, max_length=255)

    @model_validator(mode="after")
    def _non_zero(self):
        if self.amount_krw == 0:
            raise ValueError("amount_krw must be non-zero")
        return self


class GoalContributionRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    amount_krw: Decimal
    contributed_on: date
    note: str | None
    created_at: datetime


class GoalForecast(BaseModel):
    """Every number here is recomputable by hand from the goal and its
    progress — see app/services/goal_forecast.py. `factors` is the reasoning
    behind `verdict`; never render one without the other (XAI guardrail)."""

    verdict: GoalVerdict
    progress_krw: Decimal
    progress_percent: Decimal
    days_elapsed: int
    days_total: int
    days_remaining: int
    pace_krw_per_day: Decimal
    projected_final_krw: Decimal | None
    projected_percent_of_target: Decimal | None
    # The counterfactual: the daily rate that exactly lands on target, and how
    # far the current pace is from it (positive = behaviour must change).
    required_krw_per_day: Decimal | None
    required_change_krw_per_day: Decimal | None
    factors: list[XAIFactor]


class GoalRead(BaseModel):
    id: uuid.UUID
    name: str
    goal_type: GoalType
    target_amount_krw: Decimal
    starting_amount_krw: Decimal
    category: ExpenseCategory | None
    start_date: date
    target_date: date
    created_at: datetime
    forecast: GoalForecast
